File: examples/nodejs/08-error-handling.ts
/**
 * Error Handling Example
 *
 * Demonstrates how to handle various error conditions with ModelSuite.
 *
 * Requirements:
 * - Set MISTRAL_API_KEY environment variable
 *
 * Run:
 *   npx ts-node 08-error-handling.ts
 */

import {
    JsModelSuiteClient as ModelSuiteClient,
    JsMessage as Message,
    JsCompletionRequest as CompletionRequest,
} from 'modelsuite'

async function basicErrorHandling() {
    const client = ModelSuiteClient.fromEnv()

    // Use "provider/model" format for explicit provider routing
    const request = CompletionRequest
        .create('mistral/mistral-large-latest', [Message.user('Hello!')])
        .withMaxTokens(100)

    try {
        const response = await client.complete(request)
        console.log('Success:', response.textContent())
    } catch (error) {
        if (error instanceof Error) {
            console.log('Error:', error.message)
        }
    }
}

async function handleSpecificErrors() {
    const client = ModelSuiteClient.fromEnv()

    const request = CompletionRequest
        .create('mistral/mistral-large-latest', [Message.user('Hello!')])
        .withMaxTokens(100)

    try {
        const response = await client.complete(request)
        console.log('Success:', response.textContent())
    } catch (error) {
        if (error instanceof Error) {
            const message = error.message

            if (message.includes('Authentication')) {
                console.log('Authentication failed')
                console.log('Please check your API key and try again.')
            } else if (message.includes('RateLimit')) {
                console.log('Rate limited')
                console.log('Too many requests. Please wait and try again.')
            } else if (message.includes('ContextLength')) {
                console.log('Context too long')
                console.log('Try reducing your input or using a model with larger context.')
            } else if (message.includes('InvalidRequest')) {
                console.log('Invalid request:', message)
                console.log('Check your request parameters.')
            } else if (message.includes('ProviderNotFound')) {
                console.log('Provider not found')
                console.log('Make sure the provider is configured with an API key.')
            } else if (message.includes('Timeout')) {
                console.log('Request timed out')
                console.log('The request took too long. Try again or reduce complexity.')
            } else {
                console.log('Error:', message)
            }
        }
    }
}

async function demonstrateProviderNotFound() {
    const client = ModelSuiteClient.fromEnv()

    try {
        // Try to use a provider that's not configured
        await client.completeWithProvider(
            'definitely_not_a_real_provider',
            CompletionRequest
                .create('some-model', [Message.user('Hello!')])
                .withMaxTokens(100)
        )
    } catch (error) {
        if (error instanceof Error) {
            console.log('Expected error:', error.message)
            console.log('Available providers:', client.providers())
        }
    }
}

async function retryOnRateLimit() {
    const client = ModelSuiteClient.fromEnv()

    const request = CompletionRequest
        .create('mistral/mistral-large-latest', [Message.user('Hello!')])
        .withMaxTokens(100)

    const maxRetries = 3
    let retryDelay = 1000  // Start with 1 second

    for (let attempt = 0; attempt < maxRetries; attempt++) {
        try {
            const response = await client.complete(request)
            console.log(`Success on attempt ${attempt + 1}`)
            console.log('Response:', response.textContent())
            return response
        } catch (error) {
            if (error instanceof Error && error.message.includes('RateLimit')) {
                if (attempt < maxRetries - 1) {
                    console.log(`Rate limited. Waiting ${retryDelay}ms before retry...`)
                    await new Promise(resolve => setTimeout(resolve, retryDelay))
                    retryDelay *= 2  // Exponential backoff
                } else {
                    console.log('Max retries exceeded')
                    throw error
                }
            } else {
                throw error
            }
        }
    }
}

async function safeComplete(
    client: InstanceType<typeof ModelSuiteClient>,
    request: InstanceType<typeof CompletionRequest>,
    defaultResponse: string = 'Unable to generate response'
): Promise<string> {
    try {
        const response = await client.complete(request)
        return response.textContent()
    } catch (error) {
        if (error instanceof Error) {
            const message = error.message

            if (message.includes('Authentication')) {
                return 'Error: Invalid API key'
            }
            if (message.includes('RateLimit')) {
                return 'Error: Too many requests, please try again later'
            }
            if (message.includes('ContextLength')) {
                return 'Error: Input too long'
            }
            if (message.includes('InvalidRequest')) {
                return `Error: Invalid request - ${message}`
            }
            if (message.includes('ProviderNotFound')) {
                return 'Error: Provider not available'
            }
            if (message.includes('Timeout')) {
                return 'Error: Request timed out'
            }
            return `Error: ${message}`
        }
        return defaultResponse
    }
}

async function main() {
    console.log('='.repeat(50))
    console.log('Example 1: Basic Error Handling')
    console.log('='.repeat(50))
    await basicErrorHandling()

    console.log('\n' + '='.repeat(50))
    console.log('Example 2: Specific Error Types')
    console.log('='.repeat(50))
    await handleSpecificErrors()

    console.log('\n' + '='.repeat(50))
    console.log('Example 3: Provider Not Found')
    console.log('='.repeat(50))
    await demonstrateProviderNotFound()

    console.log('\n' + '='.repeat(50))
    console.log('Example 4: Safe Completion Wrapper')
    console.log('='.repeat(50))
    const client = ModelSuiteClient.fromEnv()
    const result = await safeComplete(
        client,
        CompletionRequest
            .create('mistral/mistral-large-latest', [Message.user('Say hello briefly')])
            .withMaxTokens(50)
    )
    console.log('Safe result:', result)
}

main().catch(console.error)

File: examples/python/08_error_handling.py
"""
Error Handling Example

Demonstrates how to handle various error conditions with ModelSuite.

Requirements:
- Set MISTRAL_API_KEY environment variable

Run:
    python 08_error_handling.py
"""

from modelsuite import (
    ModelSuiteClient,
    Message,
    CompletionRequest,
    # Error types
    ModelSuiteError,
    ProviderNotFoundError,
    AuthenticationError,
    RateLimitError,
    InvalidRequestError,
    ContextLengthError,
    TimeoutError,
)


def basic_error_handling():
    """Basic try/except pattern for ModelSuite errors."""
    client = ModelSuiteClient.from_env()

    # Use "provider/model" format for explicit provider routing
    request = CompletionRequest(
        model="mistral/mistral-large-latest",
        messages=[Message.user("Hello!")],
        max_tokens=100,
    )

    try:
        response = client.complete(request)
        print(f"Success: {response.text_content()}")
    except ModelSuiteError as e:
        print(f"ModelSuite error: {e}")


def handle_specific_errors():
    """Handle specific error types differently."""
    client = ModelSuiteClient.from_env()

    request = CompletionRequest(
        model="mistral/mistral-large-latest",
        messages=[Message.user("Hello!")],
        max_tokens=100,
    )

    try:
        response = client.complete(request)
        print(f"Success: {response.text_content()}")

    except AuthenticationError as e:
        # Invalid or expired API key
        print(f"Authentication failed: {e}")
        print("Please check your API key and try again.")

    except RateLimitError as e:
        # Too many requests
        print(f"Rate limited: {e}")
        # Check if retry_after is available
        if hasattr(e, "retry_after_seconds") and e.retry_after_seconds:
            print(f"Retry after {e.retry_after_seconds} seconds")

    except ContextLengthError as e:
        # Input too long for model
        print(f"Context too long: {e}")
        print("Try reducing your input or using a model with larger context.")

    except InvalidRequestError as e:
        # Bad request parameters
        print(f"Invalid request: {e}")
        print("Check your request parameters.")

    except ProviderNotFoundError as e:
        # Provider not configured
        print(f"Provider not found: {e}")
        print("Make sure the provider is configured with an API key.")

    except TimeoutError as e:
        # Request timed out
        print(f"Request timed out: {e}")
        print("The request took too long. Try again or reduce complexity.")

    except ModelSuiteError as e:
        # Catch-all for other ModelSuite errors
        print(f"ModelSuite error: {e}")


def demonstrate_provider_not_found():
    """Show what happens with unconfigured provider."""
    client = ModelSuiteClient.from_env()

    try:
        # Try to use a provider that's not configured
        response = client.complete_with_provider(
            "definitely_not_a_real_provider",
            CompletionRequest(
                model="some-model",
                messages=[Message.user("Hello!")],
                max_tokens=100,
            ),
        )
    except ProviderNotFoundError as e:
        print(f"Expected error: {e}")
        print(f"Available providers: {client.providers()}")


def demonstrate_invalid_request():
    """Show what happens with invalid parameters."""
    client = ModelSuiteClient.from_env()

    try:
        # Invalid max_tokens (negative)
        response = client.complete(
            CompletionRequest(
                model="mistral/mistral-large-latest",
                messages=[Message.user("Hello!")],
                max_tokens=-1,  # Invalid!
            )
        )
    except (InvalidRequestError, ModelSuiteError) as e:
        print(f"Invalid request error: {e}")


def retry_on_rate_limit():
    """Implement retry logic for rate limits."""
    import time

    client = ModelSuiteClient.from_env()

    request = CompletionRequest(
        model="mistral/mistral-large-latest",
        messages=[Message.user("Hello!")],
        max_tokens=100,
    )

    max_retries = 3
    retry_delay = 1  # Start with 1 second

    for attempt in range(max_retries):
        try:
            response = client.complete(request)
            print(f"Success on attempt {attempt + 1}")
            print(f"Response: {response.text_content()}")
            return response

        except RateLimitError as e:
            if attempt < max_retries - 1:
                wait_time = getattr(e, "retry_after_seconds", retry_delay)
                print(f"Rate limited. Waiting {wait_time}s before retry...")
                time.sleep(wait_time)
                retry_delay *= 2  # Exponential backoff
            else:
                print("Max retries exceeded")
                raise

        except ModelSuiteError as e:
            print(f"Error: {e}")
            raise


def safe_complete(client, request, default_response="Unable to generate response"):
    """Wrapper function that never raises exceptions."""
    try:
        response = client.complete(request)
        return response.text_content()
    except AuthenticationError:
        return "Error: Invalid API key"
    except RateLimitError:
        return "Error: Too many requests, please try again later"
    except ContextLengthError:
        return "Error: Input too long"
    except InvalidRequestError as e:
        return f"Error: Invalid request - {e}"
    except ProviderNotFoundError:
        return "Error: Provider not available"
    except TimeoutError:
        return "Error: Request timed out"
    except ModelSuiteError as e:
        return f"Error: {e}"
    except Exception as e:
        return f"Unexpected error: {e}"


def main():
    print("=" * 50)
    print("Example 1: Basic Error Handling")
    print("=" * 50)
    basic_error_handling()

    print("\n" + "=" * 50)
    print("Example 2: Specific Error Types")
    print("=" * 50)
    handle_specific_errors()

    print("\n" + "=" * 50)
    print("Example 3: Provider Not Found")
    print("=" * 50)
    demonstrate_provider_not_found()

    print("\n" + "=" * 50)
    print("Example 4: Safe Completion Wrapper")
    print("=" * 50)
    client = ModelSuiteClient.from_env()
    result = safe_complete(
        client,
        CompletionRequest(
            model="mistral/mistral-large-latest",
            messages=[Message.user("Say hello briefly")],
            max_tokens=50,
        ),
    )
    print(f"Safe result: {result}")


if __name__ == "__main__":
    main()
